File: Problem Set 1/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long cardNumber = get_long("Number: ");
    string cardType(long cardNumber);
    bool isValid(long cardNumber);
    
    bool cardIsValid = isValid(cardNumber);
    string card = cardType(cardNumber);

    if (cardIsValid)
    {
        printf("%s", card);
    } 
    else 
    {
        printf("INVALID\n");
    }
}

string cardType(long cardNumber)
{  
    int visa[] = {4};
    int amex[] = {34, 37};
    int masterCard[] = {51, 52, 53, 54, 55};
    
    while (cardNumber >= 100)
    {
        cardNumber /= 10;
    }

    if (cardNumber / 10 == visa[0])
    {
        return "VISA\n";
    }
    else if (cardNumber == amex[0] || cardNumber == amex[1])
    {
        return "AMEX\n";
    }
    else
    {
        for (int i = 0; i < 5; i++)
        {
            if (cardNumber == masterCard[i])
            {
                return "MASTERCARD\n";
            }
        }     
        return "INVALID\n";
    }  
}

bool isValid(long cardNumber)
{
    long digit[20] = {};
    int multipliedDigit[20] = {};
    int digitArrayCounter = 0;
    int productArrayCounter = 0;
    int sumProductDigits = 0;
    int finalSum = 0;

    do 
    {
        digit[digitArrayCounter] = cardNumber % 10;
        cardNumber /= 10;
        digitArrayCounter++;
    }
    while (cardNumber != 0);
    
    if (digitArrayCounter < 13 || digitArrayCounter > 16)
    {
        return false;
    }
    
    for (int i = 0; i < digitArrayCounter; i++)
    {
        if (i % 2 != 0)
        {
            multipliedDigit[productArrayCounter] = digit[i] * 2;

            if (multipliedDigit[productArrayCounter] < 10)
            {
                sumProductDigits += multipliedDigit[productArrayCounter];
            }
            else
            {
                sumProductDigits += multipliedDigit[productArrayCounter] / 10 + multipliedDigit[productArrayCounter] % 10;
            }
            
            productArrayCounter++;
        }
        else
        {
            finalSum += digit[i];    
        }
    }
        
    finalSum += sumProductDigits;
    
    if (finalSum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

File: Problem Set 6/credit/credit.py
from cs50 import get_int


def main():
    cardNumber = get_int("Number: ")
    
    cardIsValid = isValid(cardNumber)
    card = cardType(cardNumber)
    
    if cardIsValid:
        print(card)
    else:
        print("INVALID")


# Function which checks the card type based on the first 1 or two digits
def cardType(cardNumber: int) -> str:
    visa = [4]
    amex = [34, 37]
    masterCard = [51, 52, 53, 54, 55]
    
    while cardNumber >= 100:
        cardNumber //= 10
    
    if cardNumber // 10 in visa:
        return "VISA"
    elif cardNumber in amex:
        return "AMEX"
    elif cardNumber in masterCard:
        return "MASTERCARD"
    else:
        return "INVALID"


# Function which checks if the card is valid based on Luhn's algorithm
def isValid(cardNumber: int) -> bool:
    finalSum = 0
    sumProductDigits = 0
    digitList = []
    digitListCounter = 0
    
    # Using a while loop to add all card number digits in a list, in reverse order
    while True:
        if cardNumber == 0:
            break
        digitList.append(cardNumber % 10)
        cardNumber //= 10
        digitListCounter += 1

    # If the number of digits is not 13, 14, or 15 -> the card is invalid
    if digitListCounter not in [13, 14, 15, 16]:
        return False
    
    # Looping over each digit in the list to make the necessary calculations
    for index, digit in enumerate(digitList):
        if index % 2 != 0:
            multipliedDigit = digit * 2
            if multipliedDigit < 10:
                sumProductDigits += multipliedDigit
            else:
                sumProductDigits += multipliedDigit // 10 + multipliedDigit % 10
        else:
            finalSum += digit
            
    finalSum += sumProductDigits
    
    if finalSum % 10 == 0:
        return True
    else:
        return False


if __name__ == "__main__":
    main()